File: Client/Forms/Plugins.vb
Public Class Plugins

    Private Plugins() As Versidyne.Plugins.Services.AvailablePlugin
    Private ObjHost As Versidyne.Plugins.Host

    Private Sub Plugins_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        RetreivePlugins()

    End Sub

    Private Sub RetreivePlugins()

        Plugins = Versidyne.Plugins.Services.FindPlugins(IO.Path.GetDirectoryName(Application.ExecutablePath), "Versidyne.Plugins.Plugin")

        If Plugins Is Nothing Then

            ListView1.Items.Clear()
            ListView1.Items.Add("No plugins available")

        Else

            AddDescriptions()

            ObjHost = New PluginHost

        End If

    End Sub

    Private Sub AddDescriptions()

        Dim ObjPlugin As Versidyne.Plugins.Plugin
        Dim IntIndex As Int64

        'Loop through available plugins, creating instances and adding them to listbox
        For IntIndex = 0 To Plugins.Length - 1

            ObjPlugin = DirectCast(Versidyne.Plugins.Services.CreateInstance(Plugins(IntIndex)), Versidyne.Plugins.Plugin)
            ListView1.Items.Add(ObjPlugin.Name, 0)

        Next

        'ListBox1.SelectedIndex = 0
        ListView1.SelectedItems.Clear()

    End Sub

    Private Sub Enable()

        Dim ObjPlugin As Versidyne.Plugins.Plugin
        Dim ReturnCode As Integer
        Dim Args As Array = Nothing

        'Retreive Selected Data
        Dim SelPlugins As ListView.SelectedListViewItemCollection
        Dim SelPlugin As ListViewItem

        SelPlugins = ListView1.SelectedItems

        For Each SelPlugin In SelPlugins

            'Create and initialize plugin
            ObjPlugin = DirectCast(Versidyne.Plugins.Services.CreateInstance(Plugins(SelPlugin.Index)), Versidyne.Plugins.Plugin)
            ObjPlugin.Initialize(ObjHost)

            'Begin threading here, since this is where blocking should start

            'Run calculation and return result
            Try

                ReturnCode = ObjPlugin.Main(Args)

                If ReturnCode > 0 Then

                    MessageBox.Show("Plugin received return code: " & ReturnCode, Application.ProductName, MessageBoxButtons.OK, MessageBoxIcon.Exclamation)

                End If

            Catch

                MessageBox.Show("Plugin failed.", Application.ProductName, MessageBoxButtons.OK, MessageBoxIcon.Exclamation)

                Exit Sub

            End Try

            'End threading here

        Next

    End Sub

    Private Sub Disable()

        'End particular thread

    End Sub

    Private Sub EnableToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EnableToolStripMenuItem.Click

        Enable()

    End Sub

    Private Sub DisableToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles DisableToolStripMenuItem.Click

        Disable()

    End Sub

End Class

File: Client/Forms/Plugins.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Xml.Linq;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports


namespace Vexis
{
	public partial class Plugins
	{
		public Plugins()
		{
			InitializeComponent();
		}
		
		private Versidyne.Plugins.Services.AvailablePlugin[] Plugins_Renamed;
		private Versidyne.Plugins.Host ObjHost;
		
		public void Plugins_Load(System.Object sender, System.EventArgs e)
		{
			
			RetreivePlugins();
			
		}
		
		private void RetreivePlugins()
		{
			
			Plugins_Renamed = Versidyne.Plugins.Services.FindPlugins(System.IO.Path.GetDirectoryName(Application.ExecutablePath), "Versidyne.Plugins.Plugin");
			
			if (Plugins_Renamed == null)
			{
				
				ListView1.Items.Clear();
				ListView1.Items.Add("No plugins available");
				
			}
			else
			{
				
				AddDescriptions();
				
				ObjHost = new PluginHost();
				
			}
			
		}
		
		private void AddDescriptions()
		{
			
			Versidyne.Plugins.Plugin ObjPlugin = default(Versidyne.Plugins.Plugin);
			long IntIndex = default(long);
			
			//Loop through available plugins, creating instances and adding them to listbox
			for (IntIndex = 0; IntIndex <= Plugins_Renamed.Length - 1; IntIndex++)
			{
				
				ObjPlugin = (Versidyne.Plugins.Plugin) (Versidyne.Plugins.Services.CreateInstance(Plugins_Renamed[IntIndex]));
				ListView1.Items.Add(ObjPlugin.Name, 0);
				
			}
			
			//ListBox1.SelectedIndex = 0
			ListView1.SelectedItems.Clear();
			
		}
		
		private void Enable()
		{
			
			Versidyne.Plugins.Plugin ObjPlugin = default(Versidyne.Plugins.Plugin);
			int ReturnCode = default(int);
			Array Args = (Array) null;
			
			//Retreive Selected Data
			ListView.SelectedListViewItemCollection SelPlugins = default(ListView.SelectedListViewItemCollection);
			ListViewItem SelPlugin = default(ListViewItem);
			
			SelPlugins = ListView1.SelectedItems;
			
			foreach (ListViewItem tempLoopVar_SelPlugin in SelPlugins)
			{
				SelPlugin = tempLoopVar_SelPlugin;
				
				//Create and initialize plugin
				ObjPlugin = (Versidyne.Plugins.Plugin) (Versidyne.Plugins.Services.CreateInstance(Plugins_Renamed[SelPlugin.Index]));
				ObjPlugin.Initialize(ObjHost);
				
				//Begin threading here, since this is where blocking should start
				
				//Run calculation and return result
				try
				{
					
					ReturnCode = ObjPlugin.Main(Args);
					
					if (ReturnCode > 0)
					{
						
						MessageBox.Show("Plugin received return code: " + ReturnCode.ToString(), Application.ProductName, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
						
					}
					
				}
				catch
				{
					
					MessageBox.Show("Plugin failed.", Application.ProductName, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
					
					return;
					
				}
				
				//End threading here
				
			}
			
		}
		
		private void Disable()
		{
			
			//End particular thread
			
		}
		
		public void EnableToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			
			Enable();
			
		}
		
		public void DisableToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			
			Disable();
			
		}
		
	}
}
